src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -333,8 +337,63 @@ public String executor() {
         }
 
         @Override
-        public void messageReceived(ShardActiveRequest request, TransportChannel channel) throws Exception {
-            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+        public void messageReceived(final ShardActiveRequest request, final TransportChannel channel) throws Exception {
+            // make sure shard is really there before register cluster state obsever
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
```

Comment:
There is two tasks: get the shard and check if it is active. I changed that now to have two explicit methods for this to avoid code duplication, see 01e40f8d024e7965ab8914755f183a269908eaaa But I do not understand how the AsyncShardOperationAction#retry pattern would help here because I do not think we actually want to recurse?